File: app/src/main/java/com/andb/apps/todo/settings/SettingsActivity.java
package com.andb.apps.todo.settings;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Rect;
import android.os.Bundle;
import android.preference.PreferenceActivity;
import android.preference.PreferenceManager;
import android.util.Log;
import android.widget.ImageView;
import androidx.appcompat.widget.Toolbar;

import com.andb.apps.todo.R;
import com.andrognito.flashbar.Flashbar;
import com.jaredrummler.android.colorpicker.ColorPanelView;
import com.jaredrummler.android.colorpicker.ColorPickerDialogListener;

import org.jetbrains.annotations.NotNull;
import org.joda.time.DateTime;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import de.Maxr1998.modernpreferences.PreferenceScreen;
import de.Maxr1998.modernpreferences.PreferencesAdapter;
import me.saket.inboxrecyclerview.PullCollapsibleActivity;

/**
 * A {@link PreferenceActivity} that presents a set of application settings. On
 * handset devices, settings are presented as a single list. On tablets,
 * settings are split by category, with category headers shown to the left of
 * the list of settings.
 * <p>
 * See <a href="http://developer.android.com/design/patterns/settings.html">
 * Android Design: Settings</a> for design guidelines and the <a
 * href="http://developer.android.com/guide/topics/ui/settings.html">Settings
 * API Guide</a> for more information on developing a Settings UI.
 */
public class SettingsActivity extends PullCollapsibleActivity implements PreferencesAdapter.OnScreenChangeListener, ColorPickerDialogListener {

    /**
     * A preference value change listener that updates the preference's summary
     * to reflect its new value.
     */

    public static boolean darkTheme;
    public static boolean coloredToolbar;
    public static int themeColor;
    public static boolean folderMode;
    public static boolean subFilter;
    public static boolean subtaskDefaultShow = true;
    public static int defaultSort;
    public static DateTime timeToNotifyForDateOnly;

    private static Flashbar restartAppFlashbar;
    private static Flashbar restartAppFlashbar2;

    private static ImageView iconForeground;

    private static PreferencesAdapter preferencesAdapter = new PreferencesAdapter();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.setTheme(R.style.AppThemeLightCollapse);


        setContentView(R.layout.activity_settings);
        expandFromTop();
        Toolbar toolbar = findViewById(R.id.settings_toolbar);
        toolbar.setNavigationIcon(R.drawable.ic_clear_black_24dp);
        toolbar.inflateMenu(R.menu.toolbar_settings);
        RecyclerView prefView = findViewById(R.id.settings_rv);
        prefView.setLayoutManager(new LinearLayoutManager(this));
        prefView.setAdapter(preferencesAdapter);

        PreferenceScreen preferenceScreen = SettingsLayout.INSTANCE.createRootScreen(this);

        preferencesAdapter.setRootScreen(preferenceScreen);


        restartAppFlashbar = restartApp();
        restartAppFlashbar2 = restartApp();
    }

    @Override
    public void onScreenChanged(@NotNull PreferenceScreen preferenceScreen, boolean b) {

    }

    private Flashbar restartApp() {
        return new Flashbar.Builder(this)
                .gravity(Flashbar.Gravity.BOTTOM)
                .title("Restart App")
                .message("Restart the app for these changes to take place")
                .negativeActionText("Restart later".toUpperCase())
                .positiveActionText("Restart Now".toUpperCase())
                .negativeActionTapListener(new Flashbar.OnActionTapListener() {
                    @Override
                    public void onActionTapped(@NotNull Flashbar flashbar) {
                        flashbar.dismiss();
                    }
                })
                .positiveActionTapListener(new Flashbar.OnActionTapListener() {
                    @Override
                    public void onActionTapped(@NotNull Flashbar flashbar) {
                        Intent i = getBaseContext().getPackageManager()
                                .getLaunchIntentForPackage(getBaseContext().getPackageName());
                        i.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
                        startActivity(i);
                    }
                })
                .dismissOnTapOutside()
                .backgroundColor(SettingsActivity.themeColor)
                .build();

    }

    public static void setTimeToNotifyForDateOnly(DateTime dateTime, Context context) {
        timeToNotifyForDateOnly = dateTime;
        PreferenceManager.getDefaultSharedPreferences(context)
                .edit()
                .putLong("pref_notif_only_date", dateTime.getMillis())
                .apply();
        preferencesAdapter.getCurrentScreen().get("pref_notif_only_date").setSummary(timeToNotifyForDateOnly.toString("h:mm"));
        preferencesAdapter.notifyDataSetChanged();
    }

    @Override
    public void onDialogDismissed(int dialogId) {
    }

    @Override
    public void onColorSelected(int dialogId, int color) {
        SettingsActivity.themeColor = color;
        preferencesAdapter.getCurrentScreen().get("theme_color").commitInt(color);
        ((ColorPanelView)findViewById(R.id.colorPanel)).setColor(SettingsActivity.themeColor);
        preferencesAdapter.notifyDataSetChanged();
    }

    @Override
    public void onBackPressed() {
        if(preferencesAdapter.isInSubScreen()){
            preferencesAdapter.goBack();
        }else {
            super.onBackPressed();
        }
    }
}

File: app/src/main/java/com/andb/apps/todo/settings/SettingsActivity.kt
package com.andb.apps.todo.settings

import android.content.Context
import android.content.Intent
import android.content.res.Resources
import android.graphics.Color
import android.graphics.PorterDuff
import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.os.Bundle
import androidx.appcompat.widget.Toolbar
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.andb.apps.todo.Onboarding
import com.andb.apps.todo.R
import com.andb.apps.todo.SORT_TIME
import com.andb.apps.todo.utilities.Utilities
import com.jaredrummler.cyanea.CyaneaResources
import com.jaredrummler.cyanea.app.BaseCyaneaActivity
import com.jaredrummler.cyanea.delegate.CyaneaDelegate
import com.jaredrummler.cyanea.prefs.CyaneaSettingsActivity
import com.pixplicity.easyprefs.library.Prefs
import de.Maxr1998.modernpreferences.PreferenceScreen
import de.Maxr1998.modernpreferences.PreferencesAdapter
import me.saket.inboxrecyclerview.PullCollapsibleActivity
import org.joda.time.DateTime


class SettingsActivity : PullCollapsibleActivity(), PreferencesAdapter.OnScreenChangeListener,
    BaseCyaneaActivity {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        this.setTheme(R.style.AppThemeLightCollapse)
        setContentView(R.layout.activity_settings)

        val bg: CoordinatorLayout = findViewById(R.id.settingsCoordinator)
        bg.setBackgroundColor(cyanea.backgroundColor)

        if (intent.hasExtra("expandRect")) {
            val expandRect: Rect? = Rect.unflattenFromString(intent.extras.getString("expandRect"))
            if (expandRect != null) {
                expandFrom(expandRect)
            } else {
                expandFromTop()
            }
        } else {
            expandFromTop()
        }

        val toolbar = findViewById<Toolbar>(R.id.settings_toolbar)
        var icon: Drawable = getResources().getDrawable(R.drawable.ic_clear_black_24dp)
        icon.setColorFilter(if (Utilities.lightOnBackground(cyanea.backgroundColor)) Utilities.colorWithAlpha(Color.WHITE, 1f) else Utilities.colorWithAlpha(Color.BLACK, 1f), PorterDuff.Mode.SRC_ATOP)
        toolbar.setNavigationIcon(icon)
        toolbar.inflateMenu(R.menu.toolbar_settings)

        val prefView = findViewById<RecyclerView>(R.id.settings_rv)
        prefView.layoutManager = LinearLayoutManager(this)
        prefView.adapter = preferencesAdapter

        val preferenceScreen = SettingsLayout.createRootScreen(this)
        val themeIntent = Intent(this, CyaneaSettingsActivity::class.java)
        val tutorialIntent = Intent(this, Onboarding::class.java)
        preferencesAdapter.onScreenChangeListener = object :
            PreferencesAdapter.OnScreenChangeListener {
            override fun onScreenChanged(preferenceScreen: PreferenceScreen, b: Boolean) {
                when (preferenceScreen.title) {
                    "Theme" -> {
                        startActivity(themeIntent)
                        preferencesAdapter.goBack()
                    }
                    "Show Tutorial" -> {
                        startActivity(tutorialIntent)
                        preferencesAdapter.goBack()
                    }
                }
            }
        }

        preferencesAdapter.setRootScreen(preferenceScreen)

    }

    override fun onScreenChanged(preferenceScreen: PreferenceScreen, b: Boolean) {}


    override fun onBackPressed() {
        if (preferencesAdapter.isInSubScreen()) {
            preferencesAdapter.goBack()
        } else {
            super.onBackPressed()
        }
    }

    companion object {

        /**
         * A preference value change listener that updates the preference's summary
         * to reflect its new value.
         */


        @JvmStatic
        var coloredToolbar: Boolean = false
        @JvmStatic
        var folderMode: Boolean = false
        @JvmStatic
        var subFilter: Boolean = false
        @JvmStatic
        var subtaskDefaultShow = false
        @JvmStatic
        var defaultSort: Int = SORT_TIME

        @JvmStatic
        var timeToNotifyForDateOnly: DateTime = DateTime().withTime(8, 0, 0, 0)


        private val preferencesAdapter = PreferencesAdapter()

        fun setTimeToNotifyForDateOnly(dateTime: DateTime, context: Context) {
            timeToNotifyForDateOnly = dateTime
            Prefs.putLong("pref_notif_only_date", dateTime.millis)
            preferencesAdapter.currentScreen["pref_notif_only_date"]?.summary = timeToNotifyForDateOnly.toString("h:mm")
            preferencesAdapter.notifyDataSetChanged()
        }
    }

    private val delegate: CyaneaDelegate by lazy {
        CyaneaDelegate.create(this, cyanea, getThemeResId())
    }

    private val resources: CyaneaResources by lazy {
        CyaneaResources(super.getResources(), cyanea)
    }

    override fun attachBaseContext(newBase: Context) {
        super.attachBaseContext(delegate.wrap(newBase))
    }

    override fun getResources(): Resources = resources


}


